saya/core/src/prover/atlantic/mod.rs

```diff
@@ -0,0 +1,195 @@
+use std::{io::Write, time::Duration};
+
+use anyhow::Result;
+use cairo_vm::vm::runners::cairo_pie::CairoPie;
+use log::{debug, info, trace};
+use swiftness::TransformTo;
+use swiftness_stark::types::StarkProof;
+use tokio::sync::mpsc::{Receiver, Sender};
+use zip::{write::FileOptions, ZipWriter};
+
+use crate::{
+    block_ingestor::NewBlock,
+    prover::{Proof, Prover, ProverBuilder},
+    service::{Daemon, FinishHandle, ShutdownHandle},
+};
+
+mod client;
+use client::{AtlanticClient, AtlanticQueryStatus};
+
+const PROOF_STATUS_POLL_INTERVAL: Duration = Duration::from_secs(10);
+
+/// Prover implementation as a client to the hosted [Atlantic Prover](https://atlanticprover.com/)
+/// service.
+#[derive(Debug)]
+pub struct AtlanticProver {
+    client: AtlanticClient,
+    block_channel: Receiver<NewBlock>,
+    proof_channel: Sender<Proof>,
+    finish_handle: FinishHandle,
+}
+
+#[derive(Debug)]
+pub struct AtlanticProverBuilder {
+    api_key: String,
+    block_channel: Option<Receiver<NewBlock>>,
+    proof_channel: Option<Sender<Proof>>,
+}
+
+impl AtlanticProver {
+    async fn run(mut self) {
+        // TODO: split this type further into sub-services to allow parallelization
+        // TODO: add persistence for in-flight proof requests to be able to resume progress
```

Comment:
What would be the path for that?

Is the storage expected to only be used by the orchestrator, hence adding a new channel to notify about jobs being registered could make sense?
Like so the orchestrator can re-feed the same block in case this is necessary?

Reply:
I think the best way is to localize storage to specific services that request it. The orchestrator would not know whether any of its child services use storage or not - it'd be completely internal to them.

Then for example when the prover receives a certain block number, it can check internally whether it has already submitted the request for the block from a previous run. If so, and the block content matches, it skips submission and fetches the proof from that previous run instead. This process is opaque to any external components.